### ISSUE TITLE
Double masking implementation into structure pruning methods

### ISSUE THREAD

**Issue by silentspring2** [acada-sjtu/EdgeTraining#37]
我们从[Dynamic Sparse Graph](https://openreview.net/forum?id=H1goBoR9F7)中了解到：__在非结构化稀疏中，可以采用double masking的方法，将output feature map的稀疏性，迁移到BN层，从而避免BN层破坏掉稀疏性。__

而本设计主要target结构化稀疏。上述的double masking方法能否被迁移到结构化稀疏性上？是否会带来额外的accuracy degradation？

可以参考[summary](https://github.com/acada-sjtu/EdgeTraining/blob/master/Doc/Weekly-Report/Fuyuan%20Lyu-week%2022.md#prof-yuan-xies-paper)和[第三方CUDA实现](https://github.com/mtcrawshaw/dynamic-sparse-graph)

**Comment by silentspring2** [comment]
Please discuss @xuxiaoqiao @cqchu @rogercmq @ACA-Lab-SJTU

**Comment by ACA-Lab-SJTU** [comment]
double masking处理BN的方式，在TVM的表达范围内是否容易实现？

-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by ACA-Lab-SJTU** [comment]
另外，我一直没有理解open review中有人提到他们这种稀疏方式导致行会和不同列相乘从而难以在GPU上实现。哪位同学能详细解释一下？

-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by silentspring2** [comment]
> double masking处理BN的方式，在TVM的表达范围内是否容易实现？

依据之前的架构设计，在System level实现问题不大

**Comment by ACA-Lab-SJTU** [comment]
其控制粒度是什么？按照刚才的解释应该是Filter，这也不细呀？

另外，他们保留Neuron，舍弃计算这个思想我是认同的，毕竟要保留网络继续迭代更新的能力，我记得之前在多次讨论中更大家都说过。因此，我们目前的算法如何考虑这一点的？
-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by silentspring2** [comment]
控制粒度为output feature map中的单个元素，并非filter。

保留neuron在System层面我们已经有设计，kernel会根据input feature map不同的index来选择性计算部分kernel。

算法层面 @rogercmq

**Comment by rogercmq** [comment]
我对谢老师文章的理解：dynamic sparse graph 方法是 非结构化剪枝 的方法（感谢吕师兄指点，之前我认为是结构化的，其实是没看明白论文），考虑的是向量（权重矩阵的某一行）与向量（输入的某一列）的内积，并不是从矩阵乘法的角度出发的。所以GPU加速可能出现问题。

中科院须成忠老师做了一篇类似的工作(ICLR 2019, 题目是 Dynamic Channel Pruning: Feature Boosting and Suppression, FBS)，之前的讨论中也反复提到过这篇文章，只不过 FBS 方法是用神经网络去预测哪个channel应该计算。

两篇工作的套路类似，都是选择卷积层中的一部分“重要的” input channel, 只让 important input channel 做接下来的 CONV/FC + BN + Activation。

不同之处：

1. 相比较 FBS 还得训练一个神经网络 to predict important input channel，谢源老师的 DSG 用 random projection 降维做内积 更粗暴简单。

2. FBS是加速inference的(不能加速training的原因在于需要预训练模型训练额外的predictor)，而DSG更简单可以直接加速training。

但是我现在还没有思路，应该如何follow他们的工作提出我们自己的稀疏训练算法。

**Comment by rogercmq** [comment]
之前我和李东岳的想法是直接follow FBS 这篇文章，设计一种训练方法 让 original network 和 predictor network 同时训练，但是这种思路并不够solid。

**Comment by rogercmq** [comment]
根据学长的要求，我要做的实验是：对于三个模型：
    
    1. baseline model
    2. structured sparsity model
    3. structured sparsity model with double masking (ours)
 
    看三个指标：

    1. acc
    2. 收敛速度，以epoch和iter为单位
    3. model稳定性（robustness）

目前在 network slimming 上做了一些实验，baseline model 已获得，但是卡在了 step 2，即如何训练一个 structured sparsity model，还没有实现 step 3，即 structured sparsity model with double masking.

# Network Slimming 存在的问题:

1. 这种加速 inference 的方法需要fine-tune, 但是在我们的场景下sparse network是不断稀疏的所以没有微调这种说法。

2. 如果直接拿来做 training phase 的网络压缩, 我目前跑的实验是每20个回合减一次, 然后用剪枝后的比较稀疏的模型继续训练, 反复迭代整个过程直到模型收敛。

3. 但是这种方法（pruning for inference）是有问题的：

    - 具体在哪个 epoch 去剪枝是empirical and case-by-case 的，也就是说我现在跑的这个network slimming 并不符合我们动态剪枝的要求，我必须在代码里指明哪个epoch作为breakpoint去做剪枝。
    - 如果确定在某 epoch 剪枝，全局剪枝率(在 network slimming 中是)作为超参，如果剪枝率太大则 accuracy drop 惨不忍睹，剪枝率太小则收敛时间太长。稀疏化训练的 robustness 就会有问题。例如：

        > i. After training 20 epochs, VGG11 model accuracy on CIFAR10: 87.6%
        > ii. Pruning... 
        > iii. After pruning (剪枝率0.2), VGG11 model accuracy on CIFAR10: **70.4%**
        > ...
    - 压缩后的模型不能继续用之前 optimizer 对应的值 (weight decay, learning rate, etc.) 更新参数，因为网络架构改变了，原来炼丹的超参不一定是最好的选择。

# 目前的工作: 放弃network slimming

1. 复现 Learning Structured Sparsity in Deep Neural Networks 

    > NIPS 2016, Yiran Chen Group
    > 最早提出在训练过程中根据 group LASSO 结构化稀疏的一篇文章
    > 源码是 Caffe 但是
    > Github Link: [Caffe Source Code](https://github.com/wenwei202/caffe/tree/scnn)
    > **正在做的工作**: re-implement by pyTorch

**Comment by silentspring2** [comment]
Caffe and PyTorch can convert models with each other via ONNX. It might save some times during reimplementation process. At least, you may not need to write the network structure again.
